contrib/globalsign/mgo/mgo.go

```diff
@@ -0,0 +1,395 @@
+package mgo
+
+import (
+	"context"
+
+	"github.com/globalsign/mgo"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+type mongoConfig struct {
+	serviceName string
+}
+
+// MongoOption represents an option that can be passed to Dial
+type MongoOption func(*mongoConfig)
+
+func defaults(cfg *mongoConfig) {
+	cfg.serviceName = "mongodb"
+}
+
+// WithServiceName sets the service name for a given MongoDB context.
+func WithServiceName(name string) MongoOption {
+	return func(cfg *mongoConfig) {
+		cfg.serviceName = name
+	}
+}
+
+// Dial opens a connection to a MongoDB server and ties it to the
+// given context for tracing MongoDB calls.
+func Dial(ctx context.Context, url string, opts ...MongoOption) (*WrapSession, error) {
```

Comment:
So one pattern I've taken to doing is to put the context in the config, then creating a package-level `WithContext` which would return a `MongoOption`: https://github.com/DataDog/dd-trace-go/blob/b4b3c577c5f2e9487d7474b9e4f3f7b65c1a95c8/contrib/syndtr/goleveldb/leveldb/option.go

I think doing it that way we mostly could preserve the original `Dial` function signature (since variadic arguments are optional)

Reply:
Ooh, that's a good idea.  I'll update.